File: src/utils/api/tools/network.ts
/**
 * 工具模块 - 网络工具（延迟测试 / 服务器状态 / TCP 检测 / 端口列表）
 *
 * 对应后端 `tools_manager` 的 network_latency_test / server_ping / tcp_check /
 * list_open_ports action。
 */

import { TOOLS_ACTIONS, toolsManager } from './core'

// ==================== 网络延迟测试 ====================

/** 延迟测试条目 */
export interface LatencyItem {
  url: string
  latency_ms: number | null
  status_code: number
  error: string
}

/** 延迟测试结果 */
export interface NetworkLatencyResult {
  results: LatencyItem[]
}

/** 网络延迟测试 */
export function networkLatencyTest(urls: string[]): Promise<NetworkLatencyResult> {
  return toolsManager<NetworkLatencyResult>(TOOLS_ACTIONS.NETWORK_LATENCY_TEST, { urls })
}

// ==================== 服务器状态检测 ====================

/** 服务器状态检测结果 */
export interface ServerPingResult {
  motd: string
  /** 原始 MOTD（保留 § 格式化代码，供前端解析为彩色显示） */
  motd_raw: string
  online: number
  max: number
  version: string
  latency_ms: number
  favicon: string | null
  error: string
}

/** 服务器状态检测（SLP 协议） */
export function serverPing(host: string, port: number): Promise<ServerPingResult> {
  return toolsManager<ServerPingResult>(TOOLS_ACTIONS.SERVER_PING, { host, port })
}

/** TCP 端口连通性检测结果 */
export interface TcpCheckResult {
  /** 是否可连接 */
  reachable: boolean
  /** TCP 握手耗时（毫秒），失败时为 0 */
  latency_ms: number
  /** 失败原因（成功时为空） */
  error: string
}

/**
 * TCP 端口连通性检测（仅三次握手，3 秒超时）
 *
 * 用于 Frp 等非 Minecraft 协议服务的端口可达性检查，
 * 与 `serverPing`（SLP 协议）的区别：不发送应用层数据，适用于任意 TCP 服务。
 */
export function tcpCheck(host: string, port: number): Promise<TcpCheckResult> {
  return toolsManager<TcpCheckResult>(TOOLS_ACTIONS.TCP_CHECK, { host, port })
}

// ==================== 地址延迟测试 ====================

/** 地址延迟测试目标 */
export interface AddressTarget {
  /** 显示名（如「南京」），缺省用 host */
  name?: string
  host: string
  port: number
  /** 测延迟协议：tcp（默认，TCP 握手）/ udp（UDP 探针）/ ping（ICMP，系统 ping） */
  protocol?: 'tcp' | 'udp' | 'ping'
}

/** 地址延迟测试单条结果 */
export interface AddressLatencyItem {
  name: string | null
  host: string
  port: number
  protocol: string
  /** 是否可达 */
  reachable: boolean
  /** 延迟（毫秒），失败时为 0 */
  latency_ms: number
  /** 失败原因（成功时为空） */
  error: string
}

/** 地址延迟测试结果 */
export interface AddressLatencyResult {
  results: AddressLatencyItem[]
  /** 持续测试任务 id（persistent=true 时返回，供停止） */
  task_id: string | null
}

/** 地址延迟持续测试 emit 事件名（payload = AddressLatencyResult） */
export const LATENCY_UPDATE_EVENT = 'tools-latency-update'

/**
 * 地址延迟测试（tcp 握手 / udp 探针 / 系统 ping）
 * @param persistent true 时后端按 intervalMs 周期测试并经 `tools-latency-update` 事件推送，需调用 addressLatencyStop 停止
 */
export function addressLatencyTest(
  targets: AddressTarget[],
  opts?: { persistent?: boolean; intervalMs?: number },
): Promise<AddressLatencyResult> {
  return toolsManager<AddressLatencyResult>(TOOLS_ACTIONS.ADDRESS_LATENCY_TEST, {
    targets,
    persistent: opts?.persistent ?? false,
    interval_ms: opts?.intervalMs ?? 3000,
  })
}

/** 停止持续地址延迟测试 */
export function addressLatencyStop(): Promise<Record<string, never>> {
  return toolsManager<Record<string, never>>(TOOLS_ACTIONS.ADDRESS_LATENCY_STOP)
}

/** 本机监听端口条目 */
export interface OpenPortInfo {
  local_addr: string
  port: number
  protocol: string
  process_name: string | null
  pid: number | null
}

/** 列出本机监听端口结果 */
export interface ListOpenPortsResult {
  ports: OpenPortInfo[]
}

/** 列出本机所有监听中的 TCP/UDP 端口（供 Frp 内网端口选择） */
export function listOpenPorts(): Promise<ListOpenPortsResult> {
  return toolsManager<ListOpenPortsResult>(TOOLS_ACTIONS.LIST_OPEN_PORTS)
}

// ==================== 正版玩家皮肤 ====================

/** 正版玩家皮肤获取结果 */
export interface SkinFetchResult {
  /** 玩家名（正版 API 返回的规范化名称） */
  name: string
  /** 玩家 UUID（32 位十六进制，无连字符） */
  uuid: string
  /** 皮肤模型："slim"（Alex 细手臂）| "classic"（Steve 粗手臂） */
  skin_model: string
  /** 皮肤图片地址 */
  skin_url: string
  /** 皮肤 PNG（base64 data URI，供直接预览） */
  skin_image: string
  /** 披风地址（无披风时为 null） */
  cape_url: string | null
  /** 披风 PNG（base64 data URI，无披风时为 null） */
  cape_image: string | null
  /** 失败原因（成功时为空） */
  error: string
}

/** 获取正版玩家皮肤（输入玩家名，返回 UUID / 模型 / 皮肤与披风图片） */
export function skinFetch(name: string): Promise<SkinFetchResult> {
  return toolsManager<SkinFetchResult>(TOOLS_ACTIONS.SKIN_FETCH, { name })
}

/**
 * 保存皮肤图片到本地路径（base64 → PNG 文件）
 * @param savePath 保存路径（含文件名，如 D:/skin/Steve.png）
 * @param imageBase64 图片 base64（不含 data URI 前缀）
 */
export function skinSaveImage(savePath: string, imageBase64: string): Promise<{ success: boolean }> {
  return toolsManager<{ success: boolean }>(TOOLS_ACTIONS.SKIN_SAVE_IMAGE, {
    save_path: savePath,
    image_base64: imageBase64,
  })
}

File: src-tauri/src/commands/tools/types/network.rs
use serde::{Deserialize, Serialize};

/// 网络延迟测试请求参数
#[derive(Debug, Serialize, Deserialize)]
pub struct NetworkLatencyTestParams {
    /// 待测 URL 列表
    pub urls: Vec<String>,
}

/// 网络延迟测试结果
#[derive(Debug, Serialize, Deserialize)]
pub struct NetworkLatencyResult {
    pub results: Vec<LatencyItem>,
}

/// 单个 URL 的延迟测试条目
#[derive(Debug, Serialize, Deserialize)]
pub struct LatencyItem {
    pub url: String,
    /// 延迟（毫秒），失败时为 None
    pub latency_ms: Option<u64>,
    /// HTTP 状态码（如 200），失败时为 0
    pub status_code: u16,
    /// 失败原因（成功时为空）
    pub error: String,
}

/// 服务器状态检测请求参数
#[derive(Debug, Serialize, Deserialize)]
pub struct ServerPingParams {
    pub host: String,
    pub port: u16,
}

/// 服务器状态检测结果
#[derive(Debug, Serialize, Deserialize)]
pub struct ServerPingResult {
    /// 服务器 MOTD（纯文本，已从 JSON/section 符号中提取）
    pub motd: String,
    /// 服务器 MOTD 原始文本（保留 § 格式化代码，供前端解析为彩色显示）
    pub motd_raw: String,
    /// 当前在线人数
    pub online: i32,
    /// 最大人数
    pub max: i32,
    /// 服务器版本（如 "1.20.4"）
    pub version: String,
    /// 延迟（毫秒）
    pub latency_ms: u64,
    /// Favicon（base64 data URI），无则为 None
    pub favicon: Option<String>,
    /// 失败原因（成功时为空）
    pub error: String,
}

/// TCP 端口连通性检测请求参数
///
/// 用于 Frp 等非 Minecraft 协议服务的端口可达性检查：
/// 仅做 TCP 三次握手，不发送任何应用层数据，3 秒超时。
/// 与 `ServerPingParams`（SLP 协议）的区别：本检测不依赖应用层协议，
/// 适用于 frps / 数据库 / 任意 TCP 服务。
#[derive(Debug, Serialize, Deserialize)]
pub struct TcpCheckParams {
    pub host: String,
    pub port: u16,
}

/// TCP 端口连通性检测结果
#[derive(Debug, Serialize, Deserialize)]
pub struct TcpCheckResult {
    /// 是否可连接
    pub reachable: bool,
    /// TCP 握手耗时（毫秒），失败时为 0
    pub latency_ms: u64,
    /// 失败原因（成功时为空）
    pub error: String,
}

/// 地址延迟测试目标
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AddressTarget {
    /// 显示名（如「南京」），缺省用 host
    #[serde(default)]
    pub name: Option<String>,
    /// 目标主机（域名或 IP）
    pub host: String,
    /// 目标端口（ping 协议忽略）
    pub port: u16,
    /// 测延迟协议：tcp（默认，TCP 握手）/ udp（UDP 探针）/ ping（ICMP，系统 ping）
    #[serde(default = "default_latency_protocol")]
    pub protocol: String,
}

fn default_latency_protocol() -> String {
    "tcp".to_string()
}

/// 地址延迟测试请求参数
#[derive(Debug, Serialize, Deserialize)]
pub struct AddressLatencyTestParams {
    /// 待测目标列表
    pub targets: Vec<AddressTarget>,
}

/// 地址延迟测试单条结果
#[derive(Debug, Serialize, Deserialize)]
pub struct AddressLatencyItem {
    pub name: Option<String>,
    pub host: String,
    pub port: u16,
    pub protocol: String,
    /// 是否可达
    pub reachable: bool,
    /// 延迟（毫秒），失败时为 0
    pub latency_ms: u64,
    /// 失败原因（成功时为空）
    pub error: String,
}

/// 地址延迟测试结果
#[derive(Debug, Serialize, Deserialize)]
pub struct AddressLatencyResult {
    pub results: Vec<AddressLatencyItem>,
}

/// 本机监听端口条目
#[derive(Debug, Serialize, Deserialize)]
pub struct OpenPortInfo {
    /// 本地绑定的完整地址（如 "0.0.0.0:7000" / "127.0.0.1:25565"）
    pub local_addr: String,
    /// 端口号
    pub port: u16,
    /// 协议：tcp / udp
    pub protocol: String,
    /// 占用该端口的进程名（拿不到时为 None）
    pub process_name: Option<String>,
    /// 占用该端口的进程 PID（拿不到时为 None）
    pub pid: Option<u32>,
}

/// 列出本机监听端口结果
#[derive(Debug, Serialize, Deserialize)]
pub struct ListOpenPortsResult {
    pub ports: Vec<OpenPortInfo>,
}

/// 正版玩家皮肤获取请求参数
#[derive(Debug, Serialize, Deserialize)]
pub struct SkinFetchParams {
    /// 玩家名（不区分大小写，正版 API 会规范化为实际注册名）
    pub name: String,
}

/// 正版玩家皮肤获取结果
#[derive(Debug, Serialize, Deserialize)]
pub struct SkinFetchResult {
    /// 玩家名（正版 API 返回的规范化名称）
    pub name: String,
    /// 玩家 UUID（32 位十六进制，无连字符）
    pub uuid: String,
    /// 皮肤模型："slim"（Alex 细手臂）| "classic"（Steve 粗手臂）
    pub skin_model: String,
    /// 皮肤图片地址
    pub skin_url: String,
    /// 皮肤 PNG（base64 data URI，供前端直接预览）
    pub skin_image: String,
    /// 披风地址（无披风时为 None）
    pub cape_url: Option<String>,
    /// 披风 PNG（base64 data URI，无披风时为 None）
    pub cape_image: Option<String>,
    /// 失败原因（成功时为空）
    pub error: String,
}

/// 保存皮肤图片请求参数
#[derive(Debug, Serialize, Deserialize)]
pub struct SkinSaveImageParams {
    /// 保存路径（含文件名，如 D:/skin/Steve.png）
    pub save_path: String,
    /// 图片 base64（不含 data URI 前缀）
    pub image_base64: String,
}
